src/datachain/data_storage/sqlite.py

```diff
@@ -105,6 +105,11 @@ def from_db_file(cls, db_file: Optional[str] = None) -> "SQLiteDatabaseEngine":
             engine = sqlalchemy.create_engine(
                 "sqlite+pysqlite:///", creator=lambda: db, future=True
             )
+            # ensure we run SA on_connect init (e.g it registers regexp function),
+            # also makes sure that it's consistent. Otherwise in some cases it
+            # seems we are getting different results if engine object is used in a
+            # different thread first and enine is not used in the Main thread.
+            engine.connect()
```

Comment:
@dtulga since you are dealing with SQLite recently, could you please check this as well. I just want to make sure it doesn't affect your improvements PR.

Reply:
Unfortunately, based on my testing, this does reintroduce the `OSError: [Errno 24] Too many open files` errors when running tests locally (this doesn't happen in CI). Is there any other way to register these functions that doesn't require creating many more new connections? (Or closing these new connections once done?)

Reply:
yes, there might be a way to do this. I can take a look. I was hesitant to do this to not introduce a clash (when in some cases it registers our function in some cases SA, etc). But again, I can take a look.

since we pass `creator=lambda: db` and `db` in this case seems to be an already connected instance (see the `sqlite3.connect` above) I'm curios why would it be creating a new connection though 🤔

Reply:
I did some investigation, and using this instead seems to work for me locally (no more errors):
```
with engine.connect() as _:
    # Auto-close the connection using the context manager.
    pass
```
This is because engine.connect() returns a new connection, which is intended to be used as a context manager, to auto-close that connection when it is done being used.
See the documentation here: https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Engine.connect

Reply:
thanks, this is helpful. I then don't understand though what does `creator=lambda: db` do, tbh

https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.creator

are you reading it the same way as I do - that it should be returning the same connection (`db`)?

Reply:
I'm not entirely sure, but maybe it has something to do with how the engine still thinks the new connection "created" with `engine.connect()` is open? When the `engine.connect()` context manager "closes" the connection, it's just returned to the pool, the underlying DBAPI connection isn't actually closed (yet), but maybe having it still "open" interferes with normal closing / garbage collection? (I'm not really entirely sure, to be honest.)

Reply:
okay, I'll check the code a bit behind it! thanks @dtulga